### ISSUE TITLE
Inference with FlowPSAM+FlowISAM

### ISSUE THREAD

**Issue by jes-bro** [Jyxarthur/flowsam#8]
The paper mentions "layering the motion segmentation masks behind RGB-based
segmentation masks." I don't see any layering in the eval script. How/where do you perform inference for  FlowP-SAM + FlowI-SAM?

**Comment by charigyang** [comment]
Hi, thanks for your interest in our work.

This is L66-68 in seq_level_postprocess.py

**Comment by jes-bro** [comment]
I'm assuming pred_dir should have the RGB/FlowPSAM-based segmentations and the flow_pred_dir should have the flow/FlowISAM based ones? Thanks

**Comment by charigyang** [comment]
That is correct — we will clarify this in the code/md shortly to make this clearer

**Comment by jes-bro** [comment]
Thanks so much!

**Comment by jes-bro** [comment]
Another question: In the dataset directory structure:

{data_name}/
├── JPEGImages/
│   └── {category_name}/
│       ├── 00000.jpg
│       └── ......
├── FlowImages_gap1/
│   └── {category_name}/
│       ├── 00000.png
│       └── ......
├── ...... (More flow images)

Is it assumed that flow images gap 1 is between JPEG images 00000 and 00001 and gap 2 would be between 00000 and 00002?

**Comment by charigyang** [comment]
That's correct, more precisely flowimages_gap_1/00000.png is _from_ JPEG images 00000.jpg _to_ 00001.jpg

**Comment by charigyang** [comment]
Leaving it as `None` defaults to evaluating on the entire validation set. You may put in your list of sub-sequences instead if you're debugging something and don't want to evaluate on all sequences, but otherwise just leave it as such.

L35-52 of data/dataloaders/ytvos_loader.py should reflect this example

**Comment by jes-bro** [comment]
It looks like seq_loevel_postprocess is missing a read import- where does that come from?

**Comment by charigyang** [comment]
Hi, sorry -- we'll add to the main version soon, but meanwhile this is the function.

```
def read(file):
	assert type(file) is str, "file is not str %r" % str(file)
	assert os.path.isfile(file) is True, "file does not exist %r" % str(file)
	assert file[-4:] == '.flo', "file ending is not .flo %r" % file[-4:]
	f = open(file,'rb')
	flo_number = np.fromfile(f, np.float32, count=1)[0]
	assert flo_number == TAG_FLOAT, 'Flow number %r incorrect. Invalid .flo file' % flo_number
	w = np.fromfile(f, np.int32, count=1)
	h = np.fromfile(f, np.int32, count=1)
	data = np.fromfile(f, np.float32, count=2*w[0]*h[0])
	#data = np.fromfile(f, np.float32, count=2*w*h)
	# Reshape data into 3D array (columns, rows, bands)
	flow = np.resize(data, (int(h), int(w), 2))	
	f.close()

	return flow
```

**Comment by jes-bro** [comment]
Is seq_level_postprocess how you generated the figures on the website? As in the segmentation overlaid with the RGB?

**Comment by charigyang** [comment]
The overlay code uses the mask output together with cv2.addWeighted

**Comment by jes-bro** [comment]
My flowisam and flowpsam predictions on the DAVIS16 black swan seem noisier than the ones that you all have on your site. I ran the flow prediction script from this repo and then flowisam and flowpsam inference with evaluation.py. Is that how you all produced the segmentations for those figures or was there additional post-processing?

**Comment by jes-bro** [comment]
Oh the swan one was FlowPSAM+FlowISAM, were the segmentations noisier when you ran each method individually on that one?

**Comment by jes-bro** [comment]
Another question, in some of the figures on the website, it says flowISAM (sequential), does that mean the sequential post processing script was used to generate it? Thanks!

**Comment by charigyang** [comment]
For single-object cases we just use the first mask for all frames, no need to run. For multi-object, yes.

**Comment by jes-bro** [comment]
How do you filter out just the first mask?

**Comment by jes-bro** [comment]
Also, where does the sum function used here come from?

```
def iou(masks, gt, thres=0.5):
    """ IoU predictions """
    masks = (masks>thres).float()
    gt = (gt>thres).float()
    intersect = (masks * gt).sum(dim=[-2, -1])
    union = masks.sum(dim=[-2, -1]) + gt.sum(dim=[-2, -1]) - intersect
    empty = (union < 1e-6).float()
    iou = torch.clip(intersect/(union + 1e-12) + empty, 0., 1.)
    return iou
```

I got this error: 

```
/flowsam/utils.py", line 21, in iou
    intersect = (masks * gt).sum(dim=[-2, -1])
TypeError: _sum() got an unexpected keyword argument 'dim'
```
I also got an error about floats in iou. Is that version of the iou function up to date?

And just to double check, the flow_dir directories should point to Flows_gapX directories, right? With the .flo files? Thanks!

**Comment by jes-bro** [comment]
I think in seq_level_postprocess.py the masks are being np.stacked together, and it seems like they should be torch.stacked together, if iou assumes that masks is a tensor?

**Comment by charigyang** [comment]
```How do you filter out just the first mask?```

As in, just do f = processMultiSeg(...)[0:1] -- this picks the first one (the one with highest predicted score)

```it seems like they should be torch.stacked```

Very likely the case. If this doesn't work let me know and I'll find a fix, otherwise I'll update it shortly.

``` the flow_dir directories should point to Flows_gapX directories, right? With the .flo files? ```

Yes this is correct.

**Comment by jes-bro** [comment]
When you specify the number of objects in the eval, does that correspond to the number of masks in the segmentation?

**Comment by charigyang** [comment]
That might be in the GT seg? so single object would be 1 and multi object will be max in that dataset

**Comment by jes-bro** [comment]
```
Very likely the case. If this doesn't work let me know and I'll find a fix, otherwise I'll update it shortly.
``` 
In terms of seq_level_postprocess.py I haven't been able to get it to produce good masks. They're really pixelated for some reason. The only thing I modified was converting between numpy arrays and torch tensors to resolve errors.

**Comment by jes-bro** [comment]
Would you mind updating the sequential post processing script to match whatever it was when it worked? I would love to be able to reproduce your results. Thank you!

**Comment by charigyang** [comment]
I'll likely have time this weekend to update. 

Meanwhile could you check whether the individual masks are correct i.e. frame-wise score matches the table?

**Comment by charigyang** [comment]
I've updated the iou in utils.py -- I uploaded the wrong version initially. The emp is just a tiny consistency on how to deal with empty masks (to be consistent with previous literature).

If your masks look pixelated -- could you check your initial individual masks prediction whether they look reasonable?

**Comment by WuJuli** [comment]
I have one question: What should I set to the "gap"?
